lib/kubeclient/informer.rb

```diff
@@ -0,0 +1,85 @@
+module Kubeclient
+  # caches results for multiple consumers to share and keeps them updated with a watch
+  class Informer
+    def initialize(client, resource_name, reconcile_timeout: 15 * 60)
+      @client = client
+      @resource_name = resource_name
+      @reconcile_timeout = reconcile_timeout
+      @cache = nil
+      @started = nil
+      @watching = []
+    end
+
+    def list
+      @cache.values
+    end
+
+    def watch(&block)
+      with_watching(&block)
+    end
+
+    def watch_diff(&block)
+      with_watching(&block)
+    end
+
+    # not implicit so users know they have to `stop`
+    def start_worker
+      @worker = Thread.new do
+        loop do
+          fill_cache
+          watch_to_update_cache
+        rescue StandardError
+          # need to keep retrying since we work in the background
+        ensure
+          sleep(1) # do not overwhelm the api-server if we are somehow broken
+        end
+      end
+      sleep(0.01) until @cache
+    end
+
+    def stop_worker
+      @worker&.kill # TODO: be nicer ?
+    end
+
+    private
+
+    def with_watching
+      queue = Queue.new
+      @watching << queue
+      loop do
+        x = queue.pop
+        yield(x)
+      end
+    ensure
+      @watching.delete(queue)
+    end
+
+    def cache_key(resource)
+      resource.dig(:metadata, :selfLink)
+    end
+
+    def fill_cache
+      reply = @client.get_entities(nil, @resource_name, raw: true, resource_version: '0')
```

Comment:
- if the version was too old it would results in a restart of the list/watch loop, so we should be good (never saw this happening)
- watch does not receive the `0`, but what the list returned ... but it could end up in a replay since the list might have gotten old data ... so ideally we'd keep track of the last resourceVersion we saw and then do something with that (with a "latest" approach we'd lose some events though)